### PR TITLE
chore: add DIP-related comments

### PR DESCRIPTION
**Pull request by ntn-x2**
Fixes https://github.com/KILTprotocol/ticket/issues/2560.

- [x] ~WIP.~ To rebase on top of https://github.com/KILTprotocol/kilt-node/pull/583 once merged.

### CODE REVIEW

File: crates/kilt-dip-support/src/did.rs [hunk 1]
```diff
@@ -62,6 +73,30 @@ impl From<RevealedDidKeysSignatureAndCallVerifierError> for u8 {
 	}
 }
 
+/// Proof verifier that tries to verify a DID signature over a given payload by
+/// using one of the DID keys revealed in the Merkle proof. This verifier is to
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/did.rs:77]:
```suggestion
/// using one of the DID keys revealed in the Merkle proof. This verifier is
```

File: crates/kilt-dip-support/src/did.rs [hunk 2]
```diff
@@ -62,6 +73,30 @@ impl From<RevealedDidKeysSignatureAndCallVerifierError> for u8 {
...
+/// typically be used in conjunction with a verifier that takes a user-provided
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/did.rs:78]:
```suggestion
/// typically used in conjunction with a verifier that takes a user-provided
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 3]
```diff
@@ -118,8 +122,49 @@ where
 	}
 }
 
-// Implements the same `IdentityProvider` trait, but it is internally configured
-// by receiving the runtime definitions of both the provider and the receiver.
+/// Proof verifier configured given a specific KILT runtime implementation.
+///
+/// It is a specialization of the
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:127]:
```suggestion
/// A specialization of the
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 4]
```diff
@@ -118,8 +122,49 @@ where
...
+/// [`GenericVersionedDipChildProviderStateProofVerifier`] type, with
+/// configurations derived from the provided KILT runtime.
+///
+/// The generic types
+/// indicate the following:
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:132]:
```suggestion
/// are the following:
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 5]
```diff
@@ -118,8 +122,49 @@ where
...
+///   (relay)chain.
+/// * `KiltDipMerkleHasher`: The hashing algorithm used by the KILT parachain
+///   for the generation of the DIP identity commitment.
+/// * `LocalDidCallVerifier`: Logic to map `RuntimeCall`s to a specific DID key
+///   relationship. This information is used once the Merkle proof is verified,
+///   to filter only the revealed keys that match the provided relationship.
+/// * `MAX_REVEALED_KEYS_COUNT`: Max number of DID keys that the verifier will
+///   accept being revealed as part of the DIP identity proof.
+/// * `MAX_REVEALED_ACCOUNTS_COUNT`: Max number of linked accounts that the
+///   verifier will accept being revealed as part of the DIP identity proof.
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:145]:
```suggestion
///   verifier will accept revealed as part of the DIP identity proof.
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 6]
```diff
@@ -457,24 +509,101 @@ mod v0 {
...
 		traits::{
 			Bump, DidSignatureVerifierContext, DipCallOriginFilter, HistoricalBlockRegistry,
-			ProviderParachainStateInfo, RelayChainStorageInfo,
+			ProviderParachainStorageInfo, RelayChainStorageInfo,
 		},
 		utils::OutputOf,
 	};
 
+	/// The expected format of a cross-chain DIP identity proof when the
+	/// identity information is being bridged from a provider that is a child of
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:518]:
```suggestion
	/// identity information is bridged from a provider that is a child of
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 7]
```diff
@@ -457,24 +509,101 @@ mod v0 {
...
+	/// `ConsumerRuntime::LocalIdentityInfo` and bumps it up if the proof is
+	/// successfully verified, to prevent replay attacks. If additional logic is
+	/// to be stored under the `ConsumerRuntime::LocalIdentityInfo` entry, a
+	/// different verifier or a wrapper around this verifier must be built.
+	///
+	/// It expects the DIP proof to be a
+	/// [`VersionedChildParachainDipStateProof`], and returns
+	/// [`RevealedDidMerkleProofLeaves`] if the proof is successfully verified.
+	/// This information is then made availabe as an origin to the downstream
+	/// call being dispatched.
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:553]:
```suggestion
	/// call dispatched.
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 8]
```diff
@@ -457,24 +509,101 @@ mod v0 {
...
+	/// * `ProviderDidKeyId`: The runtime type of a DID key ID as defined by the
+	///   KILT child parachain.
+	/// * `ProviderAccountId`: The runtime type of an account ID as defined by
+	///   the KILT child parachain.
+	/// * `ProviderWeb3Name`: The runtime type of a web3name as defined by the
+	///   KILT child parachain.
+	/// * `ProviderLinkedAccountId`: The runtime type of a linked account ID as
+	///   defined by the KILT child parachain.
+	/// * `MAX_REVEALED_KEYS_COUNT`: Max number of DID keys that the verifier
+	///   will accept being revealed as part of the DIP identity proof.
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:597]:
```suggestion
	///   will accept revealed as part of the DIP identity proof.
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 9]
```diff
@@ -457,24 +509,101 @@ mod v0 {
...
+	/// * `MAX_REVEALED_ACCOUNTS_COUNT`: Max number of linked accounts that the
+	///   verifier will accept being revealed as part of the DIP identity proof.
```
Comment (ChrisChinchilla) [inline crates/kilt-dip-support/src/export/child.rs:599]:
```suggestion
	///   verifier will accept revealed as part of the DIP identity proof.
```

File: crates/kilt-dip-support/src/export/child.rs [hunk 10]
```diff
@@ -457,24 +509,101 @@ mod v0 {
...
+		/// The cross-chain DID signature.
 		did: DipMerkleProofAndDidSignature<DipMerkleProofBlindedValues, DipMerkleProofRevealedLeaf, ParentBlockHeight>,
 	}
 
+	/// Generic proof verifier for KILT-specific DIP identity proofs coming from
+	/// a child provider running one of the available KILT runtimes.
+	/// The proof verification step is performed on every request, and this
+	/// specific verifier has no knowledge of caching or storing state about the
+	/// subject. It only takes the provided
+	/// `ConsumerRuntime::LocalIdentityInfo` and bumps it up if the proof is
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/export/child.rs:544]: It means increase by one, in this case. Is it unclear?

File: runtimes/common/src/dip/README.md [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# KILT Decentralized Identity Provider (DIP) provider specification
+
+Specification of the format of a DIP identity commitment and the expected format of a DIP identity proof for cross-chain transactions using KILT identities.
+
+## V0
+
+The V0 of the KILT DIP Provider specification defines the following components:
+
+* **Identity details**: What are the pieces of a KILT identity that can be used for cross-chain transactions. V0 defines them to include the following information:
+  * All `DidKey`s stored under the subject's DID Document. For more details about how these keys are defined, please check the [KILT DID pallet](../../../../pallets/did).
```
Comment (ChrisChinchilla) [inline runtimes/common/src/dip/README.md:10]: And same here, we should probably replace all these with hard URLs when we have them.

File: runtimes/common/src/dip/README.md [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  * All the `LinkableAccountId`s the DID subject has linked to the DID via the KILT linking pallet. For more details about how on-chain linking works, please check the [KILT lookup pallet](../../../../pallets/pallet-did-lookup/).
```
Comment (ChrisChinchilla) [inline runtimes/common/src/dip/README.md:11]:
```suggestion
  * All the `LinkableAccountId`s the DID subject has linked to the DID via the KILT linking pallet. For more details about how on-chain linking works, read the [KILT lookup pallet](../../../../pallets/pallet-did-lookup/).
```

File: runtimes/common/src/dip/README.md [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  * (OPTIONAL) The web3name of the DID subject, if present. For more details about how web3names work, please check the [KILT web3name pallet](../../../../pallets/pallet-web3-names/).
```
Comment (ChrisChinchilla) [inline runtimes/common/src/dip/README.md:12]:
```suggestion
  * (OPTIONAL) The web3name of the DID subject, if present. For more details about how web3names work, read the [KILT web3name pallet](../../../../pallets/pallet-web3-names/).
```

**Review by ChrisChinchilla — CHANGES_REQUESTED** [review]
A bunch of small tweaks for grammar polish, but overall good!

**Review by ChrisChinchilla — CHANGES_REQUESTED** [review]
Addressing "bumps"

File: pallets/pallet-dip-consumer/README.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+A consumer chain is *connected* to a provider if there is a way for the consumer chain to verify state proofs about parts of the state of the provider chain.
+
+A cross-chain transaction with DIP assumes the entity submitting the transaction has already generated a cross-chain identity commitment on the provider chain, by interacting with the DIP provider pallet on the provider chain.
+With a generated identity commitment, a cross-chain transaction flow for a generic entity `A` works as follows:
+
+1. `A` generates a state proof proving the state of the identity commitment on the provider chain.
+2. `A` generates any additional information required for an identity proof to be successfully verified by the consumer runtime.
+3. `A`, using their account `AccC` on the consumer chain, calls the `dispatch_as` extrinsic by providing its identifier on the provider chain, the generated proof, and the `Call` to be dispatched on the consumer chain.
+   1. This pallet verifies if the proof is correct, if not returns an error.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-consumer/README.md:13]: Did you want to suggest

```suggestion
   1. This pallet verifies if the proof is correct, if not it returns an error.
```
?

File: pallets/pallet-dip-provider/README.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+The pallet generates two events: a `VersionedIdentityCommitted` and a `VersionedIdentityDeleted`.
+
+The `VersionedIdentityCommited` is called whenever a new commitment is stored, and contains information about the `Identifier` of the subject, the value of the commitment, and the commitment version.
+
+Similarly, the `VersionedIdentityDeleted`, is called whenever a commitment is deleted, and contains information about the `Identifier` of the subject and the version of the commitment deleted.
+
+## Calls
+
+0. `pub fn commit_identity(origin: OriginFor<T>, identifier: T::Identifier, version: Option<IdentityCommitmentVersion> ) -> DispatchResult`: Generate a new versioned commitment for the subject identified by the provided `Identifier`. If an old commitment for the same version is present, it is overridden. Hooks are called before the new commitment is stored, and optionally before the old one is replaced.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/README.md:49]: These are the encoded indices of the calls, which are 0-based. Maybe I can specify what the numbers mean.

[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply pallets/pallet-dip-provider/README.md:49]: @ntn-x2 Ah yes, I think that might be a good idea as it does just look like an odd list, but of course 0-based makes complete sense.

File: crates/kilt-dip-support/src/merkle.rs [hunk 16]
```diff
@@ -218,9 +245,30 @@ impl From<DidMerkleProofVerifierError> for u8 {
...
-/// keys in a DID Document.
-/// Can also be used on its own, without any DID signature verification.
+/// A type that verifies a DIP Merkle proof revealing some leaves representing
+/// parts of a KILT DID identity stored on the KILT chain.
+/// If cross-chain DID signatures are not required for the specific use case,
+/// this verifier can also be used on its own, without any DID signature
+/// verification.
+/// The Merkle proof is assumed to have been generated using one of the
+/// versioned identity commitment generators, as shown in the [KILT runtime
+/// definitions](../../../runtimes/common/src/dip/README.md).
```
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply crates/kilt-dip-support/src/merkle.rs:255]: @ntn-x2 I went to check how these links looked rendered and I can't seem to find the text anywhere. Do you know where it is rendered? Or if maybe it's being skipped?

**Comment by ChrisChinchilla** [comment]
@ntn-x2 OK, I'm done, so take a run at the comments :)

File: crates/kilt-dip-support/src/merkle.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/merkle.rs:255]: Not sure if this is more stable. It points to a specific branch, which can easily get outdated.

File: runtimes/common/src/dip/README.md (hunk 11 above)
Comment (ntn-x2) [reply runtimes/common/src/dip/README.md:10]: Will keep this (and the others) open as a reminder.

File: crates/kilt-dip-support/src/merkle.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply crates/kilt-dip-support/src/merkle.rs:292]: @ntn-x2 As said, we should update it to master/main when it is stable, but more important was my comment about the fact that these don't een render anyway… Any thoughts?

**Pull request merged**